numba/targets/heapq.py

```diff
@@ -0,0 +1,235 @@
+
+# A port of https://github.com/python/cpython/blob/3.7/Lib/heapq.py
+
+from __future__ import print_function, absolute_import, division
+
+import heapq as hq
+import numpy as np
+
+from numba import types
+from numba.errors import TypingError
+from numba.extending import overload, register_jitable
+
+
+@register_jitable
+def _siftdown(heap, startpos, pos):
+    newitem = heap[pos]
+
+    while pos > startpos:
+        parentpos = (pos - 1) >> 1
+        parent = heap[parentpos]
+        if newitem < parent:
+            heap[pos] = parent
+            pos = parentpos
+            continue
+        break
+
+    heap[pos] = newitem
+
+
+@register_jitable
+def _siftup(heap, pos):
+    endpos = len(heap)
+    startpos = pos
+    newitem = heap[pos]
+
+    childpos = 2 * pos + 1
+    while childpos < endpos:
+
+        rightpos = childpos + 1
+        if rightpos < endpos and not heap[childpos] < heap[rightpos]:
+            childpos = rightpos
+
+        heap[pos] = heap[childpos]
+        pos = childpos
+        childpos = 2 * pos + 1
+
+    heap[pos] = newitem
+    _siftdown(heap, startpos, pos)
+
+
+@overload(hq.heapify)
+def hq_heapify(x):
+
+    if not isinstance(x, types.List):
+        raise TypingError('heap argument must be a list')
+
+    dt = x.dtype
+    if isinstance(dt, types.Complex):
+        msg = ("'<' not supported between instances "
+               "of 'complex' and 'complex'")
+        raise TypingError(msg)
+
+    def hq_heapify_impl(x):
+        n = len(x)
+        for i in range(n // 2 - 1, -1, -1):
+            _siftup(x, i)
+
+    return hq_heapify_impl
+
+
+@register_jitable
+def _siftdown_max(heap, startpos, pos):
+    newitem = heap[pos]
+
+    while pos > startpos:
+        parentpos = (pos - 1) >> 1
+        parent = heap[parentpos]
+        if parent < newitem:
+            heap[pos] = parent
+            pos = parentpos
+            continue
+        break
+    heap[pos] = newitem
+
+
+@register_jitable
+def _siftup_max(heap, pos):
+    endpos = len(heap)
+    startpos = pos
+    newitem = heap[pos]
+
+    childpos = 2 * pos + 1
+    while childpos < endpos:
+
+        rightpos = childpos + 1
+        if rightpos < endpos and not heap[rightpos] < heap[childpos]:
+            childpos = rightpos
+
+        heap[pos] = heap[childpos]
+        pos = childpos
+        childpos = 2 * pos + 1
+
+    heap[pos] = newitem
+    _siftdown_max(heap, startpos, pos)
+
+
+@register_jitable
+def _heapify_max(x):
+    n = len(x)
+
+    for i in range(n // 2 - 1, -1, -1):
+        _siftup_max(x, i)
+
+
+@register_jitable
+def _heapreplace_max(heap, item):
+    returnitem = heap[0]
+    heap[0] = item
+    _siftup_max(heap, 0)
+    return returnitem
+
+
+@overload(hq.heappop)
+def hq_heappop(heap):
```

Comment:
I think all these overloads need to assert the types of their arguments, typically, `heap` needs to be a `types.List` and `item` something that may be coerced to the `heap` dtype? Then the usual tests? Same goes for the functions operating on iterables below.

Reply:
I went with the extremist view that item must be exactly the same type as heap.  Let me know what you think.

Reply:
This is probably safest for now, if in practice it becomes a problem loosening the restriction can be considered.